### ISSUE TITLE
issues with openresty 1.19

### ISSUE THREAD

**Issue by csmithATsquiz** [ledgetech/ledge#196]
When running the tests against openresty 1.19, there are differences in how results are handled / printed compared to openresty 1.15.

```
$ git diff docker
diff --git a/docker/tests/docker-compose.yml b/docker/tests/docker-compose.yml
index 81bf9ab..70ace98 100644
--- a/docker/tests/docker-compose.yml
+++ b/docker/tests/docker-compose.yml
@@ -5,7 +5,7 @@ services:
         image: "ledgetech/test-runner:latest"
         volumes:
             - ../../:/code
-        command: /bin/bash -c "TEST_LEDGE_REDIS_HOST=redis make coverage"
+        command: /bin/bash -c "yum -y upgrade openresty; TEST_LEDGE_REDIS_HOST=redis make test_all"
         working_dir: /code
         depends_on:
             - redis
```

I've attached logs of what happens when I run that container.
[runner.txt](https://github.com/ledgetech/ledge/files/5941190/runner.txt)

**Comment by csmithATsquiz** [comment]
Some of the more interesting failures :

```
runner_1  | #   Failed test 'TEST 8: Generate vary key - status code ok'
runner_1  | #   at /usr/share/perl5/vendor_perl/Test/Nginx/Socket.pm line 935.
runner_1  | #          got: '500'
runner_1  | #     expected: '200'
runner_1  |
runner_1  | #   Failed test 'TEST 8: Generate vary key - pattern "[error]" should not match any line in error.log but matches line "2021/02/08 21:49:04 [error] 380\#380: *1 lua entry thread aborted: runtime error: content_by_lua(nginx.conf:191):53: Vary spec not modified with noop function - multivalue spec" (req 0)'
```

```
runner_1  | #   Failed test 'TEST 5: Create redis connection - response_body - response is expected (repeated req 0, req 0)'
runner_1  | #   at /usr/share/perl5/vendor_perl/Test/Nginx/Socket.pm line 1382.
runner_1  | #          got: 'dognil
runner_1  | # '
runner_1  | #     expected: 'dog
runner_1  | # '
runner_1  |
runner_1  | #   Failed test 'TEST 9: Create qless connection - response_body - response is expected (repeated req 0, req 0)'
runner_1  | #   at /usr/share/perl5/vendor_perl/Test/Nginx/Socket.pm line 1382.
runner_1  | #          got: 'dognil
runner_1  | # '
runner_1  | #     expected: 'dog
runner_1  | # '
runner_1  | t/01-unit/ledge.t ..........................
```

```
runner_1  | #   Failed test 'TEST 6: Concurrent SUBZERO requests - response_body - response is expected (repeated req 0, req 0)'
runner_1  | #   at /usr/share/perl5/vendor_perl/Test/Nginx/Socket.pm line 1382.
runner_1  | #          got: 'OK 2
runner_1  | # OK 1
runner_1  | # '
runner_1  | #     expected: 'OK 1
runner_1  | # OK 2
runner_1  | # '
runner_1  | # Looks like you failed 1 test of 48.
runner_1  | t/02-integration/collapsed_forwarding.t ....
```

**Comment by pintsized** [comment]
Hey Chris, thanks for this.

Your fixes for the redis client make sense - I guess they are now explicitly returning a `nil` error value on success, and so we have multiple return values messing with the test comparisons.

The vary key one looks a bit odd though. I'll have to spin this up to take a closer look, but you're getting a 500 just because the test uses `assert`. That is, it's probably something subtle in reality but we need to inspect the log for that single test case to see what's happening. You can pass `TEST_FILE=t/01-unit/cache-key-t` to run just that file, and then remove all the tests after `TEST 8`. That way, `t/servroot/logs` will be left intact and you can see what happened more clearly, or add additional logging.

I'll try to look at it myself in the coming days though.

The `Concurrent SUBZERO requests` one *may* be a false negative. IIRC those tests are reliant on timing and it's hard to make it fully deterministic, so once in a while it will fail. I might be wrong, but if it disappears then that will be why.

**Comment by hamishforbes** [comment]
Adding `--- ONLY` to the test is easier than removing the other tests :)
A bunch of them are probably just that the hash table ordering has changed, that often happens on version bumps.
The tests should really be re-written to avoid depending on order

**Comment by csmithATsquiz** [comment]
I've done a potential fix for the easy tests (all written in lua). The `t/01-unit/processor_1_0.t` tests are a little more difficult, suggestions welcome ;)

**Comment by pintsized** [comment]
Resolved in #197
